### PR TITLE
Downsampling

### PR DESCRIPTION
**Pull request by salah-zaiem**
Code for the best technique in the paper "Fine-tuning Strategies for Faster Inference using Speech Self-Supervised Models: A Comparative Study"  : https://arxiv.org/abs/2303.06740, allowing for sequence downsampling during fine-tuning of SSL models. This leads to lower inference times with low performance drops.

### CODE REVIEW

File: speechbrain/lobes/downsampling.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
+import torch
```
Comment (TParcollet) [inline speechbrain/lobes/downsampling.py:1]: Maybe a little header explaining what this class is about?

File: speechbrain/lobes/downsampling.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+import torchaudio.transforms as T
+from speechbrain.nnet.CNN import Conv1d
+from speechbrain.nnet.pooling import Pooling1d
+
+
+class Downsampler(torch.nn.Module):
+    def forward(self, x):
```
Comment (TParcollet) [inline speechbrain/lobes/downsampling.py:8]: Docstring is needed with example for testing.

File: speechbrain/lobes/downsampling.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+        return self.downsampler(x)
+
+
+class SignalDownsampler(Downsampler):
```
Comment (TParcollet) [inline speechbrain/lobes/downsampling.py:12]: Docstring is needed with example for testing.

File: speechbrain/lobes/downsampling.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    def __init__(self, downsampling_factor, initial_sampling_rate):
+        super().__init__()
+        self.downsampling_factor = downsampling_factor
+        self.target_ds_rate = int(initial_sampling_rate / downsampling_factor)
+        self.downsampler = T.Resample(
+            initial_sampling_rate, self.target_ds_rate, dtype=torch.float32
+        )
+
+
+class Conv1DDownsampler(Downsampler):
```
Comment (TParcollet) [inline speechbrain/lobes/downsampling.py:22]: Docstring is needed with example for testing.

File: speechbrain/lobes/downsampling.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+        self.downsampler = Conv1d(
+            stride=self.downsampling_factor,
+            padding="valid",
+            kernel_size=self.kernel_size,
+            out_channels=1,
+            input_shape=[None, None],
+        )
+
+
+class PoolingDownsampler(Downsampler):
```
Comment (TParcollet) [inline speechbrain/lobes/downsampling.py:36]: Docstring is needed with example for testing.

**Review by TParcollet — CHANGES_REQUESTED** [review]
Thanks! See my comments.

File: recipes/LibriSpeech/ASR/CTC/train_with_wav2vec.py [hunk 6]
```diff
@@ -64,6 +66,11 @@ def compute_forward(self, batch, stage):
         # Compute outputs
         p_tokens = None
         logits = self.modules.ctc_lin(x)
+        # Upsample the inputs if they have been highly downsampled
+        if self.hparams.upsampling:
```
[the remark this answers is not in the document]
Comment (salah-zaiem) [reply recipes/LibriSpeech/ASR/CTC/train_with_wav2vec.py:70]: Logits are to be upsampled not downsampled. If the downsampling factor is too high ( starting from 3 actually ) it is useful to avoid outputting a character every 60ms

File: tests/recipes/LibriSpeech.csv [hunk 7]
```diff
@@ -6,9 +6,9 @@ ASR,LibriSpeech,recipes/LibriSpeech/ASR/seq2seq/train.py,recipes/LibriSpeech/ASR
...
+ASR,LibriSpeech,recipes/LibriSpeech/ASR/CTC/train_with_wav2vec.py,recipes/LibriSpeech/ASR/CTC/hparams/downsampled/train_hf_wavlm_conv_downsampling.yaml,recipes/LibriSpeech/ASR/CTC/librispeech_prepare.py,recipes/LibriSpeech/ASR/CTC/README.md,,,--downsampling_factor 2 --data_folder=tests/samples/ASR/ --skip_prep=True --train_csv=tests/samples/annotation/ASR_train.csv --valid_csv=tests/samples/annotation/ASR_train.csv --test_csv=[tests/samples/annotation/ASR_train.csv] --number_of_epochs=10 --skip_prep=True,
+ASR,LibriSpeech,recipes/LibriSpeech/ASR/CTC/train_with_wav2vec.py,recipes/LibriSpeech/ASR/CTC/hparams/downsampled/train_hf_wavlm_average_downsampling.yaml,recipes/LibriSpeech/ASR/CTC/librispeech_prepare.py,recipes/LibriSpeech/ASR/CTC/README.md,,,--downsampling_factor 2 --data_folder=tests/samples/ASR/ --skip_prep=True --train_csv=tests/samples/annotation/ASR_train.csv --valid_csv=tests/samples/annotation/ASR_train.csv --test_csv=[tests/samples/annotation/ASR_train.csv] --number_of_epochs=10 --skip_prep=True,
```
[the remark this answers is not in the document]
Comment (salah-zaiem) [reply tests/recipes/LibriSpeech.csv:11]: I added two more tests there with different downsamplings and upsamplings, should I add more?

[the remark this answers is not in the document]
Comment (anautsch) [reply tests/recipes/LibriSpeech.csv]: Hi @salah-zaiem just add as many tests as you need to cover your contribution. Please let existing tests unchanged (unless they are in error or need updating for their goal).

File: recipes/LibriSpeech/ASR/CTC/train_with_wav2vec.py [hunk 8]
```diff
@@ -26,6 +26,7 @@
 from speechbrain.utils.distributed import run_on_main
 from hyperpyyaml import load_hyperpyyaml
 from pathlib import Path
+from pyctcdecode import build_ctcdecoder
```
Comment (Adel-Moumen) [inline recipes/LibriSpeech/ASR/CTC/train_with_wav2vec.py:29]: it should be optional nop? Now this is mandatory to pip install `pyctcdecode` in order to use the CTC wav2vec...

**Review by TParcollet — APPROVED** [review]
LGTM

File: recipes/LibriSpeech/ASR/CTC/train_with_wav2vec.py (hunk 8 above)
Comment (salah-zaiem) [reply recipes/LibriSpeech/ASR/CTC/train_with_wav2vec.py:29]: Yes it should be optional, will put the import later
